### PR TITLE
chore: Update Cloudprovider interface with driftReason 

### PR DESCRIPTION
**Pull request by engedaam**
Fixes #N/A 

**Description**
- Bumping karpetner-core version and passing drift reason to karpenter-core

**How was this change tested?**
- `make battletest`
- `/karpenter snapshot`

**Does this change impact docs?**
- [ ] Yes, PR includes docs updates
- [ ] Yes, issue opened: # 
- [x] No

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/cloudprovider/cloudprovider.go [hunk 1]
```diff
@@ -189,22 +189,22 @@ func (c *CloudProvider) Delete(ctx context.Context, machine *v1alpha5.Machine) e
 	return c.instanceProvider.Delete(ctx, id)
 }
 
-func (c *CloudProvider) IsMachineDrifted(ctx context.Context, machine *v1alpha5.Machine) (bool, error) {
+func (c *CloudProvider) IsMachineDrifted(ctx context.Context, machine *v1alpha5.Machine) (cloudprovider.DriftReason, error) {
 	// Not needed when GetInstanceTypes removes provisioner dependency
 	provisioner := &v1alpha5.Provisioner{}
 	if err := c.kubeClient.Get(ctx, types.NamespacedName{Name: machine.Labels[v1alpha5.ProvisionerNameLabelKey]}, provisioner); err != nil {
-		return false, client.IgnoreNotFound(fmt.Errorf("getting provisioner, %w", err))
+		return "", client.IgnoreNotFound(fmt.Errorf("getting provisioner, %w", err))
```
Comment (njtran) [inline pkg/cloudprovider/cloudprovider.go:196]: Can we use the disruption.NotDrifted() here? Or is there a dependency cycle?

File: pkg/cloudprovider/drift.go [hunk 2]
```diff
@@ -23,97 +23,118 @@ import (
...
+		return disruption.NotDrifted, fmt.Errorf("calculating securitygroup drift, %w", err)
 	}
 	subnetDrifted, err := c.isSubnetDrifted(instance, nodeTemplate)
 	if err != nil {
-		return false, fmt.Errorf("calculating subnet drift, %w", err)
+		return disruption.NotDrifted, fmt.Errorf("calculating subnet drift, %w", err)
 	}
-
-	return amiDrifted || securitygroupDrifted || subnetDrifted || c.areStaticFieldsDrifted(machine, nodeTemplate), nil
+	drifted := lo.FindOrElse([]cloudprovider.DriftReason{amiDrifted, securitygroupDrifted, subnetDrifted, c.areStaticFieldsDrifted(machine, nodeTemplate)}, "", func(i cloudprovider.DriftReason) bool {
```
Comment (njtran) [inline pkg/cloudprovider/drift.go:57]: Same on the enum here
```suggestion
	drifted := lo.FindOrElse([]cloudprovider.DriftReason{amiDrifted, securitygroupDrifted, subnetDrifted, c.areStaticFieldsDrifted(machine, nodeTemplate)}, disruption.NotDrifted, func(i cloudprovider.DriftReason) bool {
```

File: pkg/fake/cloudprovider.go [hunk 3]
```diff
@@ -57,14 +58,14 @@ func (c *CloudProvider) GetInstanceTypes(_ context.Context, _ *v1alpha5.Provisio
 	}, nil
 }
 
-func (c *CloudProvider) IsMachineDrifted(_ context.Context, machine *v1alpha5.Machine) (bool, error) {
+func (c *CloudProvider) IsMachineDrifted(_ context.Context, machine *v1alpha5.Machine) (cloudprovider.DriftReason, error) {
 	nodeAMI := machine.Labels[v1alpha1.LabelInstanceAMIID]
 	for _, ami := range c.ValidAMIs {
 		if nodeAMI == ami {
-			return false, nil
+			return "", nil
```
Comment (njtran) [inline pkg/fake/cloudprovider.go:65]: lastly here on `disruption.NotDrifted`

**Review by njtran — COMMENTED** [review]
Looks good, just some nits on replacing the "" with the enum.

**Review by engedaam — COMMENTED** [review]
/karpenter snapshot

File: pkg/cloudprovider/cloudprovider.go [hunk 4]
```diff
@@ -189,22 +189,22 @@ func (c *CloudProvider) Delete(ctx context.Context, machine *v1alpha5.Machine) e
...
 	if err != nil {
-		return false, client.IgnoreNotFound(fmt.Errorf("resolving node template, %w", err))
+		return "", client.IgnoreNotFound(fmt.Errorf("resolving node template, %w", err))
 	}
 	drifted, err := c.isNodeTemplateDrifted(ctx, machine, provisioner, nodeTemplate)
 	if err != nil {
-		return false, err
+		return "", err
 	}
 	return drifted, nil
```
Comment (jonathan-innis) [inline pkg/cloudprovider/cloudprovider.go:209]:
```suggestion
	return driftReason, nil
```

File: pkg/cloudprovider/cloudprovider.go [hunk 5]
```diff
@@ -189,22 +189,22 @@ func (c *CloudProvider) Delete(ctx context.Context, machine *v1alpha5.Machine) e
...
 	if provisioner.Spec.ProviderRef == nil {
-		return false, nil
+		return "", nil
 	}
 	nodeTemplate, err := c.resolveNodeTemplate(ctx, nil, provisioner.Spec.ProviderRef)
 	if err != nil {
-		return false, client.IgnoreNotFound(fmt.Errorf("resolving node template, %w", err))
+		return "", client.IgnoreNotFound(fmt.Errorf("resolving node template, %w", err))
 	}
 	drifted, err := c.isNodeTemplateDrifted(ctx, machine, provisioner, nodeTemplate)
```
Comment (jonathan-innis) [inline pkg/cloudprovider/cloudprovider.go:205]:
```suggestion
	driftReason, err := c.isNodeTemplateDrifted(ctx, machine, provisioner, nodeTemplate)
```

File: pkg/cloudprovider/suite_test.go [hunk 6]
```diff
@@ -59,6 +60,8 @@ import (
 	. "github.com/aws/karpenter-core/pkg/test/expectations"
 )
 
+const NotDrifted corecloudproivder.DriftReason = ""
```
Comment (jonathan-innis) [inline pkg/cloudprovider/suite_test.go:63]: I get why you did this but why not just do the string conversion ad-hoc

File: pkg/cloudprovider/drift.go [hunk 7]
```diff
@@ -23,97 +23,119 @@ import (
 
 	"github.com/aws/karpenter-core/pkg/apis/v1alpha5"
 	"github.com/aws/karpenter-core/pkg/cloudprovider"
+	"github.com/aws/karpenter-core/pkg/controllers/machine/disruption"
```
Comment (jonathan-innis) [inline pkg/cloudprovider/drift.go:26]: We shouldn't be importing this package here

File: pkg/cloudprovider/drift.go [hunk 8]
```diff
@@ -30,90 +30,111 @@ import (
 	"github.com/aws/karpenter/pkg/utils"
 )
 
-func (c *CloudProvider) isNodeTemplateDrifted(ctx context.Context, machine *v1alpha5.Machine, provisioner *v1alpha5.Provisioner, nodeTemplate *v1alpha1.AWSNodeTemplate) (bool, error) {
+const (
+	AMIDrift           cloudprovider.DriftReason = "AMIDrift"
+	SubnetDrift        cloudprovider.DriftReason = "SubnetDrift"
+	SecurityGroupDrift cloudprovider.DriftReason = "SecurityGroupDrift"
+	StaticDrift        cloudprovider.DriftReason = "StaticDrift"
```
Comment (jonathan-innis) [inline pkg/cloudprovider/drift.go:37]: Maybe `NodeTemplateStaticDrift`?

**Review by jonathan-innis — COMMENTED** [review]
One comment

**Review by jonathan-innis — APPROVED** [review]
LGTM 🚀

**Pull request merged**